File: runtime/rust/titan-ast-runtime/titan-ast-runtime-lib/src/derived_terminal_grammar_automata_data.rs
use std::collections::HashMap;

use crate::ast::{Grammar, Token};

#[derive(Clone)]
pub(crate) struct RootTerminalGrammarMap {
    pub(crate) root_terminal_grammar: Grammar,
    pub(crate) text_terminal_map: HashMap<Vec<u8>, Grammar>,
}

#[derive(Clone)]
pub(crate) struct DerivedTerminalGrammarAutomataData {
    pub(crate) count: i32,
    pub(crate) root_terminal_grammar_maps: Vec<RootTerminalGrammarMap>,
}

impl Default for DerivedTerminalGrammarAutomataData {
    fn default() -> Self {
        Self {
            count: 0,
            root_terminal_grammar_maps: Default::default()
        }
    }
}

File: runtime/java/titan-ast/src/titan/ast/grammar/token/DerivedTerminalGrammarAutomataData.java
package titan.ast.grammar.token;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import titan.ast.grammar.Grammar;

/**
 * 只保留依赖的数据，真正的实现在runtime里面.
 *
 * @author tian wei jun
 */
public class DerivedTerminalGrammarAutomataData {
  public List<RootTerminalGrammarMap> rootTerminalGrammarMaps = new ArrayList<>();

  public boolean isEmpty() {
    return rootTerminalGrammarMaps.isEmpty();
  }

  public int sizeOfTextTerminalMap() {
    int size = 0;
    for (RootTerminalGrammarMap rootTerminalGrammarMap : rootTerminalGrammarMaps) {
      size += rootTerminalGrammarMap.textTerminalMap.size();
    }
    return size;
  }

  public static class RootTerminalGrammarMap {
    public Grammar rootTerminalGrammar; // not null
    public HashMap<String, Grammar> textTerminalMap; // not empty

    public RootTerminalGrammarMap(
        Grammar rootTerminalGrammar, HashMap<String, Grammar> textTerminalMap) {
      this.rootTerminalGrammar = rootTerminalGrammar;
      this.textTerminalMap = textTerminalMap;
    }
  }
}
